Kernel TLS socket API

borisPis opened the pull request:
I would like to introduce you to the new kernel API for TLS transmit-side data-path available since Linux 4.13. 

Placing the data-path of TLS in the kernel has several advantages. First, an efficient TLS sendfile system call implementation is possible with such a socket. Second, data from userspace could be encrypted when passing the user-kernel boundary, reducing the overhead of existing implementations that require at least encrypting the data and then copying from user to kernel.
Third, TLS **data-path** crypto offload to hardware requires a kernel infrastructure for the TLS data-path.

To sum up, a new kernel socket option is introduced to add support for the data-path of TLS and it will soon be common in operating systems.

The user API is as follows: (also see [[1]](https://github.com/torvalds/linux/blob/master/Documentation/networking/tls.txt))

Creating a TLS connection
-------------------------
First create a new TCP socket and set the TLS ULP.

```
  sock = socket(AF_INET, SOCK_STREAM, 0);
  setsockopt(sock, SOL_TCP, TCP_ULP, "tls", sizeof("tls"));
```

Setting the TLS ULP allows us to set/get TLS socket options. 
Currently only the symmetric encryption is handled in the kernel. Patch 3 sets the TLS ULP in BIO_new_socket(int fd, int close_flag) and Patch 5 sets it in the SSL_set_fd functions.

All the parameters required to move the data-path to the kernel are available after processing the change cipher state message during the handshake.
Patch 5 adds code similar to the one below in tls1_change_cipher_state.

There is a separate socket option for moving the transmit and the receive into the kernel. Current code, supports only the transmit side, TLS1.2 and AES-GCM128. In the future, this will be extended to the receive side and TLS1.3.

Below is the code that sets the socket option which moves the transmit data-path into the kernel. Patch 1 adds a new internal header internal/ktls.h that implements the new socket options required. It also provides the necessary stub functions for fallback to the OpenSSL data-path when ktls is not supported by the kernel. It is possible to disable ktls using the no-ktls config option added in Patch 2.

```
 /* From linux/tls.h */
 struct tls_crypto_info {
         unsigned short version;
         unsigned short cipher_type;
 };

  struct tls12_crypto_info_aes_gcm_128 {
          struct tls_crypto_info info;
          unsigned char iv[TLS_CIPHER_AES_GCM_128_IV_SIZE];
          unsigned char key[TLS_CIPHER_AES_GCM_128_KEY_SIZE];
          unsigned char salt[TLS_CIPHER_AES_GCM_128_SALT_SIZE];
          unsigned char rec_seq[TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE];
  };

  struct tls12_crypto_info_aes_gcm_128 crypto_info;

  crypto_info.info.version = TLS_1_2_VERSION;
  crypto_info.info.cipher_type = TLS_CIPHER_AES_GCM_128;
  memcpy(crypto_info.iv, iv_write, TLS_CIPHER_AES_GCM_128_IV_SIZE);
  memcpy(crypto_info.rec_seq, seq_number_write, TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE);
  memcpy(crypto_info.key, cipher_key_write, TLS_CIPHER_AES_GCM_128_KEY_SIZE);
  memcpy(crypto_info.salt, implicit_iv_write, TLS_CIPHER_AES_GCM_128_SALT_SIZE);

  setsockopt(sock, SOL_TLS, TLS_TX, &crypto_info, sizeof(crypto_info));
```

Sending TLS application data
----------------------------

After setting the TLS_TX socket option all application data sent over this socket is encrypted using TLS using the parameters provided in the socket option and encapsulated in TLS application data records.

For example, an application can send an encrypted hello world application data record as follows:

```
  const char *msg = "hello world\n";
  send(sock, msg, strlen(msg));

```
send() data is directly encrypted from the userspace buffer provided to the encrypted kernel send buffer if possible.

The sendfile system call will send the file's data over TLS records of maximum length (2^14). Currently, it is not possible to control the fragment length through the API, but such an option could be added in the future.  

```
  file = open(filename, O_RDONLY);
  fstat(file, &stat);
  sendfile(sock, file, &offset, stat.st_size);
```

TLS records are created and sent after each send() call, unless MSG_MORE is passed.  MSG_MORE will delay creation of a record until MSG_MORE is not passed, or the maximum record size is reached or an alert record needs to be sent.

The kernel will need to allocate a buffer for the encrypted data.
This buffer is allocated at the time send() is called, such that either the entire send() call will return -ENOMEM (or block waiting for memory), or the encryption will always succeed.  If send() returns -ENOMEM and some data was left on the socket buffer from a previous call using MSG_MORE, the MSG_MORE data is left on the socket buffer.

Send TLS control messages
-------------------------

Other than application data, TLS has control messages such as alert messages (record type 21) and handshake messages (record type 22), etc. These messages can be sent over the socket by providing the TLS record type via a CMSG.
Patch 3 calls the following function (in ktls.h) that sends @data of @length bytes using a record of type @record_type.
```
/* send TLS control message using record_type */
  static int klts_send_ctrl_message(int sock, unsigned char record_type,
                                  void *data, size_t length)
  {
        struct msghdr msg = {0};
        int cmsg_len = sizeof(record_type);
        struct cmsghdr *cmsg;
        char buf[CMSG_SPACE(cmsg_len)];
        struct iovec msg_iov;   /* Vector of data to send/receive into.  */

        msg.msg_control = buf;
        msg.msg_controllen = sizeof(buf);
        cmsg = CMSG_FIRSTHDR(&msg);
        cmsg->cmsg_level = SOL_TLS;
        cmsg->cmsg_type = TLS_SET_RECORD_TYPE;
        cmsg->cmsg_len = CMSG_LEN(cmsg_len);
        *CMSG_DATA(cmsg) = record_type;
        msg.msg_controllen = cmsg->cmsg_len;

        msg_iov.iov_base = data;
        msg_iov.iov_len = length;
        msg.msg_iov = &msg_iov;
        msg.msg_iovlen = 1;

        return sendmsg(sock, &msg, 0);
  }
```

Control message data should be provided unencrypted, and will be encrypted by the kernel. Additionally, each control message must be preceded with a BIO_CTRL_SET_OFFLOAD_TX_CTRL_MSG that specifies the record type of the upcoming message.

When using ktls, OpenSSL doesn't encrypt the payload, thus it doesn't need to copy user provided data. Patch 5 also removes all copies from the OpenSSL data-path, as a side-effect it disables the use the OpenSSL buffer and packet management APIs.

Some performance analysis of ktls compared to OpenSSL can be found at Dave Watson's paper from netdev1.2 [[2]](https://netdevconf.org/1.2/papers/ktls.pdf). The paper shows how ktls reduces latency and improves CPU utilization. Additionally, ktls enables seamless HW offload of the data-path (like LSO) which further improves both latency and CPU utilization as you can see from my talk at netdev 2.2 [[3]](https://www.netdevconf.org/2.2/papers/pismenny-tlscrypto-talk.pdf) where we show x2.6 improvment in CPU utilization compared to vanilla OpenSSL.

At a high level, the kernel TLS ULP is a replacement for the record layer of a userspace TLS library.

References:
[1] - https://github.com/torvalds/linux/blob/master/Documentation/networking/tls.txt
[2] - https://netdevconf.org/1.2/papers/ktls.pdf
[3] - https://www.netdevconf.org/2.2/papers/pismenny-tlscrypto-talk.pdf

##### Changes from the RFC pull request
- Submitted CLAs
- Rebased changes over master
- Added a new configuration option to disable ktls
- Remove the netinet/tcp.h hack

richsalz commented:
"At a high level, the kernel TLS ULP is a replacement for the record layer of a userspace TLS library."  So then what would be needed for OpenSSL to do?

borisPis commented:
OpenSSL handles the handshake entirely (state machine, asymmetric crypto, ASN1 parsing, etc.). Additionally all control messages after the handshake must be generated by OpenSSL.

richsalz commented on crypto/bio/bss_sock.c:
```diff
@@ -64,7 +65,12 @@ BIO *BIO_new_socket(int fd, int close_flag)
         return NULL;
     BIO_set_fd(ret, fd, close_flag);
-    return ret;
+#ifndef OPENSSL_NO_KTLS
```
push an error code, don't printf in the library.

borisPis replied:
The printf helps with debugging. I'll add it under an SSL_DEBUG clause. OK?

The error code wasn't propagated to the caller on purpose. We do this because the socket is created successfully regardless of the return value from ktls_enable. However, future calls to ktls_* functions will fail if this call fails. This can happen when the library is compiled with ktls but the kernel is too old. As a result, the library creates the socket successfully, even though the socket option fails.

richsalz commented on include/internal/ktls.h:
```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 1995-2018 The OpenSSL Project Authors. All Rights Reserved.
```
copyright year is wrong.

borisPis replied:
Done.

richsalz commented on include/internal/ktls.h:
```diff
@@ -0,0 +1,119 @@
+
+
+/* Dummy functions here */
+inline int ktls_enable(int fd) {
```
open brace on a line by itself.  here and throughout.

borisPis replied:
Done.

richsalz commented on include/internal/ktls.h:
```diff
@@ -0,0 +1,119 @@
+        const void *data, size_t length)
+{
+    struct msghdr msg = {0};
+    int cmsg_len = sizeof(record_type);
```
side-note, but isn't sizeof [unsigned]char 1 by definition?

borisPis replied:
Yes. I think that using record_type here helps the reader follow the code. Would you like to replace it with:
int cmsg_len = sizeof(unsigned char);
?

richsalz replied:
That's a good point, leave it.

richsalz commented on include/internal/ktls.h:
```diff
@@ -0,0 +1,119 @@
+
+#else
+
+#include <netinet/tcp.h>
```
indent the pre-processor by a space, as in
```
# include <netinet/tcp.h>
...
```
here and throughout.

borisPis replied:
Done.

Though it seems like a lot of other code throughout the project isn't consistent about this. Should I do this for the entire patchset or only this file?

richsalz replied:
I guess it depends on how many changes there are.  It's your call.

richsalz commented on include/openssl/bio.h:
```diff
@@ -144,6 +144,10 @@ extern "C" {
 # endif
 
 # define BIO_CTRL_DGRAM_SET_PEEK_MODE      71
+# define BIO_CTRL_SET_OFFLOAD_TX               72
```
the name "offload tx" seems strange to me.  is it a kernel kind of usage?  I would see it as just offload without the tx.  here and in the next few.

borisPis replied:
No, it is not a kernel kind of usage. Maybe BIO_CTRL_SET_KTLS or BIO_CTRL_SET_KTLS_TX is a better name?
The "TX" part is related to the fact that this is directional - only the transmit side is supported at the moment. A future patchset will add the receive side as well. Should we assume that the direction is implied by the surrounding code and call this just set_ktls?

richsalz replied:
I like just KTLS.

borisPis replied:
Done. There is only one place where the direction is still required - BIO_set_ktls. For this purposes a new "is_tx" parameter is added to this macro.

richsalz commented on ssl/record/rec_layer_s3.c:
```diff
@@ -1119,27 +1151,37 @@ int ssl3_write_pending(SSL *s, int type, const unsigned char *buf, size_t len,
         clear_sys_error();
         if (s->wbio != NULL) {
             s->rwstate = SSL_WRITING;
+            if (BIO_get_offload_tx(s->wbio) && type !=
```
break this line before the &&.

borisPis replied:
Done.

richsalz commented on ssl/record/rec_layer_s3.c:
```diff
@@ -1119,27 +1151,37 @@ int ssl3_write_pending(SSL *s, int type, const unsigned char *buf, size_t len,
         || (s->rlayer.wpend_type != type)) {
         SSLfatal(s, SSL_AD_INTERNAL_ERROR, SSL_F_SSL3_WRITE_PENDING,
-                 SSL_R_BAD_WRITE_RETRY);
+                SSL_R_BAD_WRITE_RETRY);
```
you broke the indent :)

borisPis replied:
Done.

richsalz commented on ssl/record/rec_layer_s3.c:
```diff
@@ -1119,27 +1151,37 @@ int ssl3_write_pending(SSL *s, int type, const unsigned char *buf, size_t len,
-                          &(SSL3_BUFFER_get_buf(&wb[currbuf])
-                            [SSL3_BUFFER_get_offset(&wb[currbuf])]),
-                          (unsigned int)SSL3_BUFFER_get_left(&wb[currbuf]));
+                    &(SSL3_BUFFER_get_buf(&wb[currbuf])
```
broke the indent :)

borisPis replied:
Done.

richsalz commented on ssl/record/rec_layer_s3.c:
```diff
@@ -1119,27 +1151,37 @@ int ssl3_write_pending(SSL *s, int type, const unsigned char *buf, size_t len,
             if (i >= 0)
                 tmpwrit = i;
+
+            /* To prevent coalescing of control and data messages,
```
put the start-comment on a line by itself for multiline comments.

borisPis replied:
Done.

richsalz commented on ssl/record/rec_layer_s3.c:
```diff
@@ -1295,8 +1337,8 @@ int ssl3_read_bytes(SSL *s, int type, int *recvd_type, unsigned char *buf,
         for (curr_rec = 0;
-             curr_rec < num_recs && SSL3_RECORD_is_read(&rr[curr_rec]);
-             curr_rec++) ;
+                curr_rec < num_recs && SSL3_RECORD_is_read(&rr[curr_rec]);
```
please unfix this :)

borisPis replied:
Done.

richsalz commented on ssl/ssl_lib.c:
```diff
@@ -1336,6 +1337,11 @@ int SSL_set_fd(SSL *s, int fd)
     SSL_set_bio(s, bio, bio);
+#ifndef OPENSSL_NO_KTLS
+    if (ktls_enable(fd)) {
+        printf("setsockopt failed %d\n", errno);
```
push an error code, don't printf in the library.

borisPis replied:
This is the same as in crypto/bio/bss_sock.c. I'll fix it throughout the patchset according to what we agree about there.

richsalz commented on ssl/ssl_lib.c:
```diff
@@ -1355,6 +1361,11 @@ int SSL_set_wfd(SSL *s, int fd)
         SSL_set0_wbio(s, bio);
+#ifndef OPENSSL_NO_KTLS
+        if (ktls_enable(fd)) {
+            printf("setsockopt failed %d\n", errno);
```
same error-code/printf issue.

borisPis replied:
See comment in crypto/bio/bss_sock.c. I'll fix it throughout the patchset according to what we agree about there.

borisPis commented:
I'm not sure I understand why does this depend on FIPS. Please explain.

BTW, I've forced pushed all changes marked as Done and a suggestion for the crypto/bio/bss_sock.c fix.

richsalz commented:
Having printf's with ifdef debug-flag is fine.  Just no printf's in production code.  And I lost the FIPS reference, not sure what I meant or posted?  Thanks.

richsalz commented:
open/close to toggle the CLA bot.

mattcaswell commented on ssl/record/rec_layer_s3.c:
```diff
@@ -743,6 +743,16 @@ int do_ssl3_write(SSL *s, int type, const unsigned char *buf,
+         * ktls doesn't modify the buffer, but to avoid a warning we need to
+         * discard the const qualifier
+         */
+        SSL3_BUFFER_set_buf(&s->rlayer.wbuf[0], (unsigned char *)buf);
```
Won't this leak any existing buffer?

borisPis replied:
Right. I haven't thought of that. I'll add a ssl3_release_write_buffers call to release the existing buffers before we start using ktls to ensure all previous buffers are released.

borisPis replied:
Done.

mattcaswell commented on ssl/record/rec_layer_s3.c:
```diff
@@ -1133,13 +1165,24 @@ int ssl3_write_pending(SSL *s, int type, const unsigned char *buf, size_t len,
+             * such as in buffer_write, we flush the BIO
+             */
+            if (BIO_get_offload_tx(s->wbio) && type != SSL3_RT_APPLICATION_DATA)
+                (void)BIO_flush(s->wbio);
```
This could fail!

borisPis replied:
Are you concerned about non-blocking IO or some other failure?
If this is a ktls socket, then non-blocking IO sendmsg will either succeed fully or fail fully. There is no problematic case where the socket writes some partial payload.
In case of failure, we expect sock_write to be called again with the control message and any additional data of the same record type.

mattcaswell replied:
I'm thinking about non-blocking IO mainly - but other failures are also a concern I supposed. Are you saying with ktls we will never have data pending in the BIO so we don't need to worry about this (in which case why are we flushing at all)?

mattcaswell replied:
This comment seems to be outstanding

mattcaswell replied:
In #7527 you said:
> What do you mean? - AFAIU it is resolved.

I still had a question on this which was unanswered, i.e. "Are you saying with ktls we will never have data pending in the BIO so we don't need to worry about this (in which case why are we flushing at all)?"

borisPis replied:
Oh, I've missed this.

You're right this might fail and we should handle this, I'll add some code to handle this failure.

borisPis commented on crypto/bio/bss_sock.c:
```diff
@@ -64,7 +65,22 @@ BIO *BIO_new_socket(int fd, int close_flag)
+         * Thus, we don't propogate this error.
+         * However, future calls to ktls will fail.
+         */
+        int rc = ktls_enable(fd);
```
ktls_enable calls the TCP_ULP socket option. This socket option doesn't change the behaviour of the TCP socket, except changing the TCP setsockopt handler to enable the processing of SOL_TLS options. Only the TLS_TX socket option changes the send/sendmsg handlers. We set it for all sockets because it allows for transparent support of all existing applications. Another option is to call enable_ktls when handling BIO_CTRL_SET_OFFLOAD_TX.

mattcaswell replied:
If we keep it here, then at least a comment to say why it isn't a problem for non-tls sockets might be in order.

borisPis replied:
Done. I'll also add some description to include/internal/ktls.h to describe these functions.

borisPis commented on crypto/bio/bss_sock.c:
```diff
@@ -64,7 +65,22 @@ BIO *BIO_new_socket(int fd, int close_flag)
+        if (rc) {
+            printf("setsockopt failed %d\n", errno);
+        }
+#endif // SSL_DEBUG
```
Done.

borisPis commented on crypto/bio/bss_sock.c:
```diff
@@ -111,7 +127,23 @@ static int sock_write(BIO *b, const char *in, int inl)
+        ret = ktls_send_ctrl_message(b->num, record_type, in, inl);
+        if (ret >= 0) {
+            ret = inl;
+        }
```
Done.

borisPis commented on include/internal/ktls.h:
```diff
@@ -0,0 +1,123 @@
+#ifndef HEADER_INTERNAL_KTLS
+# define HEADER_INTERNAL_KTLS
+
+# include <linux/version.h>
```
Right. I've added the OPENSSL_NO_KTLS check here and in a bunch of other places. I've compiled with no-ktls and with non-linux. I hope this compiles now.

borisPis commented on include/internal/ktls.h:
```diff
@@ -0,0 +1,123 @@
+
+static ossl_inline int ktls_enable(int fd)
+{
+    return setsockopt(fd, SOL_TCP, TCP_ULP, "tls", sizeof("tls"));
```
Done.

borisPis commented on include/internal/ktls.h:
```diff
@@ -0,0 +1,123 @@
+                          struct tls12_crypto_info_aes_gcm_128 *crypto_info,
+                          size_t len)
+{
+    return setsockopt(fd, SOL_TLS, TLS_TX, crypto_info,
```
Done.

borisPis commented on include/openssl/bio.h:
```diff
@@ -174,6 +178,13 @@ extern "C" {
+ * BIO_FLAGS_OFFLOAD_TX means we are using offload with this BIO for TX.
+ * BIO_FLAGS_OFFLOAD_TX_CTRL_MSG means we are about to send a ctrl message next.
+ */
+# define BIO_FLAGS_OFFLOAD_TX          0x2000
```
Done. I think I've skipped a few when I've started working with openssl just to avoid conflicts when rebasing. Then, I forgot to align it.

borisPis commented on ssl/record/ssl3_buffer.c:
```diff
@@ -160,7 +163,8 @@ int ssl3_release_write_buffer(SSL *s)
         wb = &RECORD_LAYER_get_wbuf(&s->rlayer)[pipes - 1];
 
-        OPENSSL_free(wb->buf);
+        if (!BIO_get_offload_tx(s->wbio))
```
Are such changes valid?
I didn't consider this as a valid use-case until now. Could you give me an example to work with?
Do you mean something like BIO_push? or is it completely changing the BIO, like from using one socket to using another socket or even a buffer/mem/file BIO?

mattcaswell replied:
> or is it completely changing the BIO, like from using one socket to using another socket or even a buffer/mem/file BIO?

I'm thinking of completely changing the BIO. I wouldn't be surprised to learn that there are existing applications that do this, e.g. swapping from a socket to a mem bio mid connection.

richsalz replied:
yuck.  i don't see a problem if that fails to work.

mattcaswell replied:
> yuck. i don't see a problem if that fails to work.

I do. I think people do this.

richsalz replied:
I think they don't, so we're equal.  Unless you have evidence...  I mean, how can you swap the BIO, e.g., if there's pending data?

mattcaswell replied:
> how can you swap the BIO, e.g., if there's pending data?

Easy. If its pending in the BIO layer then you just flush it first. If it's pending in the libssl layer it doesn't matter anyway.

borisPis commented on ssl/t1_enc.c:
```diff
@@ -316,6 +322,77 @@ int tls1_change_cipher_state(SSL *s, int which)
+    }
+
+    /* ktls supports only the maximum fragment size */
+    if (ssl_get_max_send_fragment(s) != SSL3_RT_MAX_PLAIN_LENGTH)
```
Currently it is ignored and the kernel keeps sending using the maximum size. I'll added a check in SSL_ctrl.

borisPis commented on ssl/t1_enc.c:
```diff
@@ -316,6 +322,77 @@ int tls1_change_cipher_state(SSL *s, int which)
+#endif // SSL_DEBUG
+        goto skip_offload;
+    }
+    crypto_info.info.cipher_type = TLS_CIPHER_AES_GCM_128;
```
Done.

borisPis commented on ssl/t1_enc.c:
```diff
@@ -316,6 +322,77 @@ int tls1_change_cipher_state(SSL *s, int which)
+    crypto_info.info.cipher_type = TLS_CIPHER_AES_GCM_128;
+
+    /* check version is 1.2 */
+    if (s->version != TLS1_2_VERSION) {
```
Done.

borisPis commented on ssl/t1_enc.c:
```diff
@@ -316,6 +322,77 @@ int tls1_change_cipher_state(SSL *s, int which)
+            TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE);
+
+    wbio = s->wbio;
+    if (!wbio) {
```
Done. This check is just in-case it is somehow possible. Should I remove it?

borisPis commented on ssl/t1_enc.c:
```diff
@@ -316,6 +322,77 @@ int tls1_change_cipher_state(SSL *s, int which)
+    if (BIO_get_offload_tx(s->wbio))
+        goto err;
+
+    (void)BIO_flush(wbio);
```
Haven't thought about it. I think we cannot proceed to use ktls if there is pending data. I'll add a goto skip_offload here.

mattcaswell commented on ssl/t1_enc.c:
```diff
@@ -316,6 +323,73 @@ int tls1_change_cipher_state(SSL *s, int which)
+
+    wbio = s->wbio;
+    if (wbio == NULL)
+        goto skip_ktls;
```
Perhaps this should be:
````
    if(!ossl_assert(wbio != NULL)) {
        SSLfatal(...);
        goto err;
    }
````

borisPis replied:
Done.

mattcaswell commented on include/internal/ktls.h:
```diff
@@ -0,0 +1,130 @@
+
+
+/* Dummy functions here */
+ossl_inline int ktls_enable(int fd)
```
Travis failures are due to this I think:
````
ssl/t1_enc.o: In function `ktls_enable':
t1_enc.c:(.text+0x280): multiple definition of `ktls_enable'
ssl/ssl_lib.o:ssl_lib.c:(.text+0x4c0): first defined here
collect2: error: ld returned 1 exit status
make[1]: *** [libssl.so] Error 1
make: *** [all] Error 2
+/// MAKE FAILED
````
Probably this should be declared `static` otherwise it gets defined multiple times in every file that includes this header.

borisPis replied:
Done. Thanks.

mattcaswell commented:
We should have a test for this code.

mattcaswell commented on crypto/bio/bss_sock.c:
```diff
@@ -111,7 +128,23 @@ static int sock_write(BIO *b, const char *in, int inl)
+#ifdef SSL_DEBUG
+        printf("\nsending ctrl msg\n");
+#endif
+        ret = ktls_send_ctrl_message(b->num, record_type, in, inl);
```
Needs an `OPENSSL_NO_KTLS` guard. This is the cause of the Appveyor failure.

borisPis replied:
Done.

richsalz commented:
Is there a way to know if the cipher suite is compatible with what's in the kernel?  Does it just happen that the kernel rejects the socket option?

borisPis commented:
The application knows that the cipher suite is supported only if it is successful in setting the crypto_info with the TLS_TX socket option. Otherwise, the cipher suite is not supported.

Maybe we should add a GET_SUPPORED_CIPHER_SUITE socket option to address this, and avoid the need for a a syscall on each tls1_change_cipher_state.

borisPis commented:
Regarding testing. Which tests would you like to see?

Upto this moment I was using s_client, s_server and the debug prints on a recent Linux Kernel to see that things are working as expected.

mattcaswell commented:
> Regarding testing. Which tests would you like to see?

I'd like to see a test in the test subdir that does this:
1) Checks that we're on a platform that supports ktls. If not skip the test
2) Creates a TLSv1.2 connection using a suitable ciphersuite etc between two endpoints. I'd anticipate this is just a single-threaded thing where the client and server take turns reading/writing to the socket
3) Sends some data each way
4) Shuts down cleanly
5) Not sure if there is a way we could test that offload to the kernel has actually taken place?

There are plenty of examples of doing this type of thing in sslapitest.c. The main difference is that this will use a "real" socket rather than a mem bio.

dot-asm commented on Configure:
```diff
@@ -1484,6 +1485,28 @@ unless ($disabled{afalgeng}) {
+              $disabled{ktls} = "too-old-kernel";
+          }
+      } else {
+          $disabled{ktls} = "cross-compiling";
```
Is cross-compilation actually mission impossible? Besides, in the context headers are more reliable source than current uname -r...

borisPis replied:
I didn't try cross-compilation. I guess it might work. It would be better if someone who can actually test it would provide the patches.

Regarding "uname -r", I just copied it from afalg. Please let me know if you have a better solution.
I don't know what you mean by context-headers. Please explain.

dot-asm replied:
> context-headers.

No dash. It's supposed to be "in the context [of this request] headers ...". [BTW, should one add comma after context?] Anyway, what I'm trying to say is that I reckon that headers are more reliable basis for taking the decision than uname -r at the moment of config script execution. As simplest example consider scenario when you installed new kernel and headers but haven't rebooted the machine yet. Which would incidentally be equivalent of cross compilation. I mean cross compilation at the very least is nothing else but compilation with alien headers. Well, one can always argue that "nobody does that" or "if they do they are doing it wrong", but it's really just a projection of one's prior experiences, and nothing says that it's the only possible experience, or that it's most satisfying one... Of course it might turn out that what I'm trying to say is not as practical as implied, but it's not inappropriate to keep trying to push boundaries, because if we don't, we won't make as much progress...

dot-asm replied:
But on more essential note I'd say that it's more than appropriate that for *starters* it's disabled by default.

dot-asm replied:
Well, this is not *formally* right either. Earlier mentioned case of not-rebooted-my-computer-yet was a simplest example of cross-compilation. So simple that it would be totally appropriate to put it in quotes, "cross-compilation". Formally right thing to do is to call *configured* compiler with *configured* flags, and ask it to compile or pre-process something. This is most reliable assumption one can make. Idea is to hold user responsible for providing right target headers for *configured* compiler invoked with *configured* flags.

borisPis replied:
I understand that for cross-compilation the configured compiler,linker,etc. need to be called, but the ktls configuration is not responsible for that.
The only thing that is relevant for this config option is the linux headers. Your previously suggested use-case should work now. But, cross-compilers usually don't provide any special headers, let alone linux kernel headers. Moreover, I haven't seen a relevant OpenSSL tutorial that mentions include path manipulation for cross-compilation.

dot-asm replied:
> cross-compilers usually don't provide any special headers,

They totally are. Cross-compiler without target platform headers is worthless.

> let alone linux kernel headers.

Even this is not true. If target platform is Linux, then cross-compiler will have headers that are customarily found in what corresponds to Debian linux-libc-dev.

> I haven't seen a relevant OpenSSL tutorial

If you play your cards right --cross-compiler-prefix is all it takes. For example I cross-compile for ARM Linux with `Configure linux-armv4 --cross-compiler-prefix=arm-linux-gnueabi-` directly on my regular computer. And there is a separate snapshot of glibc and Linux headers [and of course libraries]. Another example is Travis CI cross-compiling for mingw...

borisPis replied:
If you were to add linux kernel headers for the cross-compiled target. Where would you add them relative to your "--cross-compiler-prefix"?

dot-asm replied:
Well, it's totally up to cross-compiler provider/vendor, so there is no common answer. But in my case ARM Linux headers can be found in /usr/arm-linux-gnueabi/include/. But once again, beauty is that you don't have to know, just call cross-compiler and pass it to somebody else. Well, unless you are in business of providing cross-compiler, but we aren't and you don't seem to be...

borisPis replied:
Would something like this meet the requirement?


    if ($target =~ m/^linux/) {
        my $usr = "/usr/$config{cross_compile_prefix}";
        chop($usr);
        if ($config{cross_compile_prefix} eq "") {
            $usr = "/usr";
        }
        my $minver = (4 << 16) + (13 << 8) + 0;
        my @verstr = split(" ",`cat $usr/include/linux/version.h | grep LINUX_VERSION_CODE`);
        print STDERR "Linux vesion $verstr[2]. KTLS minmum version $minver.\n";

        if ($verstr[2] < $minver) {
            $disabled{ktls} = "too-old-kernel";
        }
    } else {
        $disabled{ktls}  = "not-linux";
    }

dot-asm commented on ssl/t1_enc.c:
```diff
@@ -316,6 +324,76 @@ int tls1_change_cipher_state(SSL *s, int which)
+
+    /* check that cipher is AES_GCM_128 */
+    if (EVP_CIPHER_mode(c) != EVP_CIPH_GCM_MODE
+        || EVP_CIPHER_key_length(c) != TLS_CIPHER_AES_GCM_128_KEY_SIZE) {
```
You can't identify AES_GCM_128 by check EVP_CIPH_GCM_MODE and key size. I mean GCM mode is not exclusive to AES.

borisPis replied:
Done.

dot-asm commented:
I totally fail to compile this on Ubuntu 17. Inclusion of linux/tls.h triggers redefinitions of bunch of structures and at the end complains about net/tcp.h being absent. net/tcp.h is present in kernel source, but apparently is not installed as part of linux-libc-dev. Presumably for good reason as it declares kernel-only functions. Either way it's effectively uncompilable on Ubuntu, so that question also is if it gets fixed. Note that question is not how to fix it, but if/when it will be fixed elsewhere.

dot-asm commented:
For reference, as far as I can tell it takes a kernel module to be loaded, and it's privileged operation. In sense that if I, as unprivileged user, run this, and module was not loaded in advance, I won't be exercising it.

dot-asm commented:
BTW, is it correctly understood that it work only for transmission? And it's kernel limitation, not suggested code's. I also wonder if there are future plans for say aes256-gcm...

borisPis commented:
Regarding the problem of compiling on Ubuntu17. You might have a kernel that is missing this fix:
https://patchwork.ozlabs.org/patch/837675/
I think that you could take this header here and use it instead:
http://elixir.free-electrons.com/linux/latest/source/include/uapi/linux/tls.h

You are right about the kernel module. The ktls_* calls will fail when the module is not loaded. These failures will not influence the normal operation of OpenSSL.

There is WIP to add the receive side of ktls, see more details here:
https://github.com/Mellanox/tls-offload/tree/tls_rx_v0

Adding AES256-GCM shouldn't be a problem, but I don't know of any specific effort in this regard.

dot-asm commented:
> Regarding the problem of compiling on Ubuntu17. You might have a kernel that is missing this fix

You have to recognize that it's not that relevant what any specific individual such as myself is missing on personal computer or what [s]he can do about it. [As a matter of fact I did something, as I did actually compile and tested it.] Relevant question is what are general users' expectations. And they customarily expect that it just works in vanilla installations. And thing is that we can't blame them for that :-(

dot-asm commented:
> https://netdevconf.org/1.2/papers/ktls.pdf

I'd say that it gives a little bit misleading picture. Naturally Figure 4 is just deceptive :-) I mean it kind of looks "wow" , but then you notice that it's only a top of a picture. Anyway, as far as I can tell improvement is not overwhelming and in such case it becomes race of implementations. I mean if another GCM implementation performs better on another processor, then you wouldn't be able to justify the suggested alternative. [But do read on!] And it's actually case that OpenSSL GCM performs better on later processors. However, besides the [obvious] possibility of adopting another GCM implementation for kernel, one can actually wonder if just measuring transfer speed is right benchmarking method to illuminate the advantage. I would guess that overall system performance under heavy load would be more adequate, i.e. how many requests does system manage to meet "before" vs. "after". Do you have data for that?

fridex commented:
> And it's actually case that OpenSSL GCM performs better on later processors.

Based on benchmarks I did this is true, but the initial idea for doing kernel TLS was not about moving AES GCM encryption/decryption into kernel because of speed, but rather because of saving context switches (that do not showed to be that resource consuming on x86 - but that is because of architecture and available kernel optimizations for context switches, benchmarks for other architectures will vary), copy operations from kernel space to user space and vice versa and introduce possibilities to do CPU offloading. This can be especially helpful for HAProxy load balancer or TLS based VPNs (that was the use case we had at Red Hat when starting with kernel TLS - see http://www.fit.vutbr.cz/study/DP/DP.php?id=18032).

borisPis commented:
> Anyway, as far as I can tell improvement is not overwhelming and in such case it becomes race of implementations.

For an overwhelming improvment you should see what capable hardware could do once you combine it with ktls (This will work seamlessly!). Check figure 1 of https://www.netdevconf.org/2.2/papers/pismenny-tlscrypto-talk.pdf

> However, besides the [obvious] possibility of adopting another GCM implementation for kernel, one can actually wonder if just measuring transfer speed is right benchmarking method to illuminate the advantage. 

There are many improvements to be made on both userspace and kernel crypto. Assuming the best implementation will be eventually available to all. There will be a few inherent advantages to ktls:
1. Reduce the number of data copies. 
  - No need to perform copies inside OpenSSL.
  - Data is not copied from user to kernel, but rather it is encrypted from user memory to kernel memory.
2. Reduced VM page management overhead
3. Enabling seamless hardware offload for TLS crypto. Potentially removing the need for an infrastructure like afalg in OpenSSL.

> I would guess that overall system performance under heavy load would be more adequate, i.e. how many requests does system manage to meet "before" vs. "after". Do you have data for that?

No. But, we are working on an academic paper with a more detailed evaluation.

dot-asm commented:
Do note that I'm not saying that advantage is little, but rather that you ought to be able to do better illuminating it :-) As for figure 1 in pismenny-tlscrypto-talk.pdf. So one has to compare +offload and ktls, right? It's also confusing because on picture it says "64 Record Size" while text says "[specifically?] offload provided [imp]rovement for maximum sized records (16K)". Can you clarify?

borisPis commented:
> So one has to compare +offload and ktls, right?

Yes. But, this wasn't our focus when doing these measurements.

> It's also confusing because on picture it says "64 Record Size" while text says "[specifically?] offload provided [imp]rovement for maximum sized records (16K)". Can you clarify?

That's the openoffice format, I thought I got rid of that :(
The graph should read the same as the text. Using maximum sized records (16K) we were able to obtain about 2.5x compared to vanilla OpenSSL.

borisPis commented:
> I'd like to see a test in the test subdir that does this:
Checks that we're on a platform that supports ktls. If not skip the test
Creates a TLSv1.2 connection using a suitable ciphersuite etc between two endpoints. I'd anticipate this is just a single-threaded thing where the client and server take turns reading/writing to the socket
Sends some data each way
Shuts down cleanly
Not sure if there is a way we could test that offload to the kernel has actually taken place?
There are plenty of examples of doing this type of thing in sslapitest.c. The main difference is that this will use a "real" socket rather than a mem bio.

I've added a test in sslapitest which opens two loopback TCP sockets and uses ktls to send data. After verifying that data is received as expected. The TLS record sequence numbers before and after sending data are compared. The write sequence number is not updated - the kernel generates the correct sequence number. However, the read sequence number of the receive is updated, because the receive path goes through ktls.

mattcaswell commented on test/sslapitest.c:
```diff
@@ -560,6 +561,103 @@ static int execute_test_large_message(const SSL_METHOD *smeth,
+        goto end;
+
+    if (!TEST_true(BIO_get_ktls(clientssl->wbio))
+        || !TEST_true(BIO_get_ktls(serverssl->wbio)))
```
Not sure how easy it would be to do, but it would be cool if we could have one side doing KTLS and one side not.

borisPis replied:
That shouldn't be a problem. I could add a new mode to ssl_st that controls this knob. Look for SSL_MODE_NO_KTLS_TX in the new patch. The TX suffix is used to prepare the code for SSL_MODE_NO_KTLS_TX in the future. I prefer to keep the default mode as enabled to simplify integration with existing code. 

Once the new mode is added, I'll split test_ktls to cover all four possible cases - enable/disable ktls and client/server.

mattcaswell commented on test/sslapitest.c:
```diff
@@ -560,6 +561,103 @@ static int execute_test_large_message(const SSL_METHOD *smeth,
+
+    if (!TEST_true(send_recv_qery(clientssl, serverssl, cfd, sfd))
+                || !TEST_true(send_recv_qery(clientssl, serverssl, cfd, sfd))
+                || !TEST_true(send_recv_qery(clientssl, serverssl, cfd, sfd)))
```
Why do we do this 3 times?

borisPis replied:
I wanted to verify that it doesn't break after running once. I'll remove the redundant two calls.

mattcaswell commented on test/ssltestlib.c:
```diff
@@ -12,6 +12,13 @@
 #include "ssltestlib.h"
 #include "testutil.h"
+#ifndef OPENSSL_NO_KTLS
+#include "internal/ktls.h"
```
Style nit: Inside a "#if" so insert a space between "#" and "include"

borisPis replied:
Fixed.

mattcaswell commented on test/ssltestlib.c:
```diff
@@ -12,6 +12,13 @@
 #include "testutil.h"
+#ifndef OPENSSL_NO_KTLS
+#include "internal/ktls.h"
+#include "internal/sockets.h"
```
Why do we need these internal headers here?

borisPis replied:
We don't need them. I'll remove these leftovers which I added while playing around with possible implementations of the test.

borisPis commented on include/internal/ktls.h:
```diff
@@ -0,0 +1,147 @@
+ * record_type instead of the default SSL3_RT_APPLICATION_DATA.
+ * When the socket is non-blocking, then this call either returns EAGAIN or
+ * the entire record is pushed to TCP. It is  impossible to send a partial send
+ * a partial record using this control message.
```
Fixed.

borisPis commented on include/internal/ktls.h:
```diff
@@ -0,0 +1,147 @@
+ */
+static ossl_inline int ktls_start(int fd,
+                                  struct tls12_crypto_info_aes_gcm_128
+                                  *crypto_info, size_t len, int is_tx)
```
I've added it after we discussed naming with @richsalz. We agreed to use ktls instead of something like ktls_tx. 

Since the offload is directional, and the BIO is oblivious to the direction, we need to provide the direction as metadata. In the near future we plan to submit the receive side support for ktls. Then, patches for OpenSSL to support it will follow. These patches will complete the use of is_tx.

Otherwise, we could remove is_tx, and assume the direction is always tx, until the rx patches will change that.

borisPis commented on include/internal/ktls.h:
```diff
@@ -0,0 +1,147 @@
+#    endif
+
+#    ifndef TCP_ULP
+#     define TCP_ULP 31
```
New glibc has it defined in netinet/tcp.h, otherwise it is defined in linux/tcp.h.
Since netinet/tcp.h is already used in other places I prefer to use it here instead of linux/tcp.h.

borisPis commented on crypto/bio/bss_sock.c:
```diff
@@ -108,10 +126,31 @@ static int sock_read(BIO *b, char *out, int outl)
 
     clear_socket_error();
-    ret = writesocket(b->num, in, inl);
+    if (BIO_should_ktls_ctrl_msg_flag(b)) {
```
It is not mandatory, because all the BIO flag setting/checking is outside of OPENSSL_NO_KTLS. This is comfortable because there are no OPENSSL_NO_KTLS sections in ssl/record/rec_layer_s3.c.
It works because the ktls flags could be set only if ktls is enabled.

Having said that, after removing all the debug prints I've moved it inside the OPENSSL_NO_KTLS section. It is a bit ugly due to the indentation required by the else clause.

borisPis commented on crypto/bio/bss_sock.c:
```diff
@@ -64,7 +65,24 @@ BIO *BIO_new_socket(int fd, int close_flag)
+        if (rc == 0) {
+#  ifdef SSL_DEBUG
+            printf("setsockopt failed %d\n", errno);
+#  endif                        /* SSL_DEBUG */
```
Done. I'll remove all debug sections here and throughout.

borisPis commented:
Ping @mattcaswell and @richsalz . AFAIK all the changes requested were made.

richsalz commented:
FWIW, I'm not going to approve this for 1.1.1  Maybe someone else will.

jrfastab commented:
Hi, @borisPis, @richsalz it looks like this has been marked after 1.1.1 which was posted here, https://github.com/openssl/openssl/releases/tag/OpenSSL_1_1_1 are there any outstanding issues that need to be resolved before it can be merged? (Looks like the comments were mostly addressed by @borisPis ?) 

I've been running this PR locally for testing and integration with some applications (with kTLS enabled kernel)  it would be great to start pointing at master. Thanks.

richsalz commented:
It's up to the openssl project to decide on this (I'm not an official part of it any more).  I hope it goes in for the next release.

mattcaswell commented:
@borisPis - can you confirm this PR is still relevant, and if so can you rebase it?

borisPis commented:
> @borisPis - can you confirm this PR is still relevant, and if so can you rebase it?

Yes, this is still relevant. I'll rebase it.

borisPis commented:
The CI failure seems to be unrelated to the patch.

@mattcaswell can you please have another look at this.

mattcaswell commented:
> The CI failure seems to be unrelated to the patch.

Are you sure about that? The CI failure seems to be related to a mem leak in the ssl3_buffer code - which is touched by this PR.

mattcaswell commented on ssl/ssl_lib.c:
```diff
@@ -1150,7 +1151,9 @@ void SSL_free(SSL *s)
     BIO_free_all(s->wbio);
+    s->wbio = NULL;
     BIO_free_all(s->rbio);
+    s->rbio = NULL;
```
I wonder why this is necessary since we later free `s` anyway.

borisPis replied:
It's a good habit to set pointers to NULL, just in case we get a double free anything funny. I could remove it if you'd like?

mattcaswell commented on ssl/ssl_lib.c:
```diff
@@ -1341,6 +1344,23 @@ int SSL_set_fd(SSL *s, int fd)
+        int rc = ktls_enable(fd);
+
+        if (rc == 0) {
+#ifdef SSL_DEBUG
```
Yuck. I'd prefer it if we don't add any more `SSL_DEBUG` stuff. `printf` just doesn't belong in the library.

mattcaswell commented on ssl/ssl_lib.c:
```diff
@@ -1360,6 +1380,23 @@ int SSL_set_wfd(SSL *s, int fd)
+            int rc = ktls_enable(fd);
+
+            if (rc == 0) {
+#ifdef SSL_DEBUG
```
Same comment here about SSL_DEBUG

mattcaswell commented on ssl/ssl_lib.c:
```diff
@@ -2185,6 +2222,8 @@ long SSL_ctrl(SSL *s, int cmd, long larg, void *parg)
     case SSL_CTRL_SET_MAX_SEND_FRAGMENT:
         if (larg < 512 || larg > SSL3_RT_MAX_PLAIN_LENGTH)
             return 0;
+        if (s->wbio && BIO_get_ktls(s->wbio))
```
This restriction should be documented somewhere. Theoretically this could be a breaking change.

mattcaswell commented on test/sslapitest.c:
```diff
@@ -656,6 +657,178 @@ static int execute_test_large_message(const SSL_METHOD *smeth,
+    /* Create a session based on SHA-256 */
+    if (!TEST_true(create_ssl_ctx_pair(TLS_server_method(),
+                                       TLS_client_method(),
+                                       TLS1_VERSION, TLS1_2_VERSION,
```
Does KTLS support TLS1_VERSION? I thought it was TLSv1.2 only? In any case I think we're going to need to skip these tests if only TLSv1.3 is enabled.

borisPis replied:
Only TLS_1_2.
Will fix.

mattcaswell commented on test/sslapitest.c:
```diff
@@ -656,6 +657,178 @@ static int execute_test_large_message(const SSL_METHOD *smeth,
+                                       TLS1_VERSION, TLS1_2_VERSION,
+                                       &sctx, &cctx, cert, privkey))
+            || !TEST_true(SSL_CTX_set_cipher_list(cctx,
+                                                  "ECDHE-RSA-AES128-GCM-SHA256"))
```
This will fail if built with no-ec

borisPis replied:
Will fix.

mattcaswell commented on test/sslapitest.c:
```diff
@@ -656,6 +657,178 @@ static int execute_test_large_message(const SSL_METHOD *smeth,
+        goto end;
+
+    if (!cis_ktls_tx) {
+        if (!TEST_true(!BIO_get_ktls(clientssl->wbio)))
```
`TEST_false` might make more sense

mattcaswell commented on test/sslapitest.c:
```diff
@@ -656,6 +657,178 @@ static int execute_test_large_message(const SSL_METHOD *smeth,
+            TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE);
+
+    /* verify the payload */
+    if (!TEST_true(memcmp(cbuf, sbuf, sizeof(cbuf)) == 0))
```
`TEST_mem_eq` would be more appropriate here

borisPis replied:
Will fix.

mattcaswell commented on test/sslapitest.c:
```diff
@@ -656,6 +657,178 @@ static int execute_test_large_message(const SSL_METHOD *smeth,
+
+    /* ktls is used then kernel sequences are used instead of OpenSSL sequences */
+    if (clientssl->mode & SSL_MODE_NO_KTLS_TX) {
+        if (!TEST_true(memcmp(crec_wseq_before, crec_wseq_after,
```
`TEST_mem_eq` again here and other places below

borisPis replied:
Will fix

borisPis commented on crypto/bio/bss_sock.c:
```diff
@@ -151,6 +176,25 @@ static long sock_ctrl(BIO *b, int cmd, long num, void *ptr)
         break;
+    case BIO_CTRL_SET_KTLS:
+# ifndef OPENSSL_NO_KTLS
+        crypto_info = (struct tls12_crypto_info_aes_gcm_128 *)ptr;
```
I've considered two possibilities to resolve this:
1. Move BIO_CTRL_SET_KTLS inside the "#ifndef OPENSSL_NO_KTLS" clause. Here and throughout. - **This is the approach I've implemented.**
2. Use a generic structure that will be converted to "struct tls12_crypto_info_aes_gcm_128" inside include/internal/ktls.h

mattcaswell replied:
I don't see how that solves the problem. The issue is that you define `BIO_CTRL_SET_KTLS` and the related macro `BIO_set_ktls` in the public header file. As such you are declaring it as part of the public API and available for use by end user applications. However it is actually impossible for such applications to use because (as I understand it) `struct tls12_crypto_info_aes_gcm_128` is internal only so applications cannot make a sensible call to `BIO_set_ktls` anyway. In reality this seems to be an entirely internal thing, so these macros should probably not appear in the public headers at all. Probably this is true for some of the other macros too.

borisPis replied:
I've relied on the code implemented in SCTP, where Linux specific code (internal structures) appear in dgram_ctrl.

I wouldn't mind to move these definitions out of the public header. But, it wouldn't be consistent to move BIO_CTRL_*KTLS to an internal header with other BIO_CTRL_* in the public header. AFAIU, there are other internal BIO_CTRL* that are already in the public header.

mattcaswell replied:
Don't use the SCTP code as a model for how to do things!! OpenSSL has had a very long history and previously *everything* was in the public headers (there wasn't really the concept of internal headers). These days we are much stricter about such things. If it appears in the public headers then it is public API. If it isn't intended for public use then it should go into an internal header. I would put a comment in the public header saying that the BIO_CTRL numbers 72, 73, etc (or whatever the numbers are) are reserved for internal use, and then define them properly in include/internal/bio.h

borisPis replied:
Thanks for the help. I'll move the definitions to include/internal/bio.h and leave a comment on the public file.

borisPis commented on doc/man3/SSL_CTX_set_mode.pod:
```diff
@@ -105,6 +105,16 @@ Enable asynchronous processing. TLS I/O operations may indicate a retry with
 
+=item SSL_MODE_NO_KTLS_TX
+
+Disable the use of the kernel TLS egress data-path. By default kernel TLS is
```
Since many users can't use this. It is disabled by default on compilation (see Configure). If however, a user has compiled OpenSSL with enable-ktls, then I assume that it should be ON by default. This API allows fine-grained control on whether to use KTLS on a specific socket.

mattcaswell replied:
Ah, right. Ok that makes sense. In which case we should be careful about the wording here then. Its only on by default if ktls support has been enabled. We should also document the restrictions that apply if KTLS is being used, i.e. no renegotiation, no setting the max fragment size, etc.

borisPis replied:
The restrictions are not inherent, but rather temporary until they are implemented. I'll add a comment about the current status, but this may change over time.

borisPis commented on doc/man3/SSL_CTX_set_mode.pod:
```diff
@@ -105,6 +105,16 @@ Enable asynchronous processing. TLS I/O operations may indicate a retry with
+=item SSL_MODE_NO_KTLS_TX
+
+Disable the use of the kernel TLS egress data-path. By default kernel TLS is
+enabled for supported ciphers. The kernel TLS data-path implements the record
```
That's right. I'll add the bit about version to the documentation. Encrypt-than-mac is unrelated though. At the moment it is just AES-GCM TLS1.2.

mattcaswell replied:
Does that mean encrypt-then-mac *is* supported by ktls?

borisPis replied:
If we ever implement Encrypt-Than-MAC (which I highly doubt) then I think it would be supported with this API.

mattcaswell replied:
Ah ok. But right now EtM is not supported, so KTLS must be disabled if it has been negotiated. We should ensure that something is documented about this. It could be kept fairly general, e.g. "By default kernel TLS is enabled if it is supported by the negotiated ciphersuites and extensions". Incidentally if EtM is not supported then this significantly reduces the usefulness of KTLS. My expectation (not based on any real data) is that the majority of TLSv1.2 connections will use it.

borisPis replied:
Sure I'll fix the description here according to your suggestion.

borisPis commented on include/internal/ktls.h:
```diff
@@ -0,0 +1,147 @@
+ */
+static ossl_inline int ktls_enable(int fd)
+{
+    return setsockopt(fd, SOL_TCP, TCP_ULP, "tls", sizeof("tls")) ? 0 : 1;
```
It is as intended.
It is a Upper Layer Protocol that enables new setsockopt calls on the TCP socket. It is also use to load the kernel TLS module if it is not loaded. See here:
https://elixir.bootlin.com/linux/v4.19/source/net/tls/tls_main.c#L727
and
https://elixir.bootlin.com/linux/v4.19/source/net/ipv4/tcp.c#L2736

borisPis commented on include/openssl/bio.h:
```diff
@@ -190,6 +202,18 @@ void BIO_clear_flags(BIO *b, int flags);
+# define BIO_should_ktls_ctrl_msg_flag(b) \
+    BIO_test_flags(b, (BIO_FLAGS_KTLS_CTRL_MSG))
+# define BIO_clear_ktls_ctrl_msg_flag(b) \
+    BIO_clear_flags(b, (BIO_FLAGS_KTLS_CTRL_MSG))
```
I've moved them inside the #ifndef OPENSSL_NO_KTLS" clause

borisPis commented:
> > The CI failure seems to be unrelated to the patch.
> 
> Are you sure about that? The CI failure seems to be related to a mem leak in the ssl3_buffer code - which is touched by this PR.

The latest patches should fix that. The problem was related to how we release buffers when using ktls, e.g. the ssl3_buffer changes.

borisPis commented:
Ping @mattcaswell. AFAIK all the changes requested were made and all tests pass.

mattcaswell commented on ssl/t1_enc.c:
```diff
@@ -316,6 +325,64 @@ int tls1_change_cipher_state(SSL *s, int which)
+    memcpy(crypto_info.rec_seq, &s->rlayer.write_sequence,
+           TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE);
+
+    BIO_set_ktls(wbio, &crypto_info, which & SSL3_CC_WRITE);
```
I still don't see any handling of Encrypt-then-Mac in the above code. Either KTLS doesn't support it - in which case we need to skip ktls if it wasn't negotiated. Or alternatively it does support it (and doesn't support non-EtM) in which case we need to skip ktls if EtM was not negotiated. Note that its a default option so in an openssl->openssl case it will always be negotiated.

borisPis replied:
These check:
`
    /* check that cipher is AES_GCM_128 */
    if (EVP_CIPHER_nid(c) != NID_aes_128_gcm
        || EVP_CIPHER_mode(c) != EVP_CIPH_GCM_MODE
        || EVP_CIPHER_key_length(c) != TLS_CIPHER_AES_GCM_128_KEY_SIZE)
        goto skip_ktls;
`
Should verify that we are not in EtM, right?

AFAIK ETM is not relevant for AES-GCM. Otherwise, what is the meaning of ETM with AES-GCM?

mattcaswell replied:
Duh...good point.

mattcaswell commented on ssl/t1_enc.c:
```diff
@@ -316,6 +325,64 @@ int tls1_change_cipher_state(SSL *s, int which)
+        goto err;
+
+    /* ktls works with user provided buffers directly */
+    ssl3_release_write_buffer(s);
```
Should we wait for success of BIO_set_ktls() before releasing the write buffer? If the underlying BIO doesn't support it, or it fails for some other reason we have needlessly released it.

borisPis replied:
What do you mean by ~wait for success~?

It is successful if we reach `ssl3_release_write_buffer(s);`

mattcaswell replied:
BIO_set_ktls() might still fail at this point, e.g. because we're using a mem BIO, or we've compiled in the dummy KTLS code because the kernel doesn't support it or it fails for some other reason.

borisPis replied:
Should be fixed now.

mattcaswell commented on ssl/record/rec_layer_s3.c:
```diff
@@ -1066,13 +1094,16 @@ int do_ssl3_write(SSL *s, int type, const unsigned char *buf,
             goto err;
         }
 
+ mac_done:
```
Couldn't this be moved further down to just above the SSL3_BUFFER_set_left() call below? You might also have to move the SSL3_RECORD_set_type() call as well.

borisPis replied:
Yes, I've moved it now.

mattcaswell commented on ssl/record/rec_layer_s3.c:
```diff
@@ -1142,6 +1173,21 @@ int ssl3_write_pending(SSL *s, int type, const unsigned char *buf, size_t len,
+             */
+            if (BIO_get_ktls(s->wbio) && type != SSL3_RT_APPLICATION_DATA) {
+                i = BIO_flush(s->wbio);
+                if (i < 0)
```
Everywhere else we call BIO_flush() in libssl we test for <= 0

mattcaswell commented on ssl/record/ssl3_buffer.c:
```diff
@@ -117,18 +117,21 @@ int ssl3_setup_write_buffer(SSL *s, size_t numwpipes, size_t len)
-                         SSL_F_SSL3_SETUP_WRITE_BUFFER, ERR_R_MALLOC_FAILURE);
-                return 0;
-            }
+            if (!s->wbio || (s->wbio && !BIO_get_ktls(s->wbio))) {
```
There is no need to test `s->wbio` twice. Also use `s->wbio == NULL`, i.e.
`if (s->wbio == NULL || !BIO_get_ktls(s->wbio)) {`

borisPis replied:
Done.

mattcaswell commented on ssl/record/ssl3_buffer.c:
```diff
@@ -117,18 +117,21 @@ int ssl3_setup_write_buffer(SSL *s, size_t numwpipes, size_t len)
+                    return 0;
+                }
+            } else
+                p = NULL;
```
Style nit: since the main `if` block is in `{}`, so must the else block

borisPis replied:
Done.

mattcaswell commented on ssl/record/ssl3_buffer.c:
```diff
@@ -160,7 +163,9 @@ int ssl3_release_write_buffer(SSL *s)
         wb = &RECORD_LAYER_get_wbuf(&s->rlayer)[pipes - 1];
 
-        OPENSSL_free(wb->buf);
+        if (wb->buf)
```
There is no need for this test. `OPENSSL_free` can cope with a NULL input

borisPis replied:
Done

mattcaswell commented on ssl/record/ssl3_buffer.c:
```diff
@@ -160,7 +163,9 @@ int ssl3_release_write_buffer(SSL *s)
 
-        OPENSSL_free(wb->buf);
+        if (wb->buf)
+            if (!s->wbio || (s->wbio && !BIO_get_ktls(s->wbio)))
```
Same comment here as for the similar line above.

borisPis replied:
Done

mattcaswell commented on ssl/ssl_lib.c:
```diff
@@ -2185,6 +2206,10 @@ long SSL_ctrl(SSL *s, int cmd, long larg, void *parg)
         if (larg < 512 || larg > SSL3_RT_MAX_PLAIN_LENGTH)
             return 0;
+#ifndef OPENSSL_NO_KTLS
+        if (s->wbio && BIO_get_ktls(s->wbio))
```
Use `s->wbio != NULL`

borisPis replied:
Done.

mattcaswell commented on test/sslapitest.c:
```diff
@@ -656,6 +657,178 @@ static int execute_test_large_message(const SSL_METHOD *smeth,
+            TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE);
+
+    if (!TEST_true(SSL_write(clientssl, cbuf, sizeof(cbuf)) == sizeof(cbuf)))
+        goto end;
```
Since the only thing that `end` does is `return 0` I'd be tempted to replace all of these with a simple `return 0`

borisPis replied:
It would be easier to change this code if we use goto here. If it is not mandatory, I'd rather keep it this, because a follow up patch for TLS Rx support might change things here.

mattcaswell replied:
Ok

mattcaswell commented on test/sslapitest.c:
```diff
@@ -656,6 +657,178 @@ static int execute_test_large_message(const SSL_METHOD *smeth,
     return testresult;
 }
 
+#ifndef OPENSSL_NO_KTLS
```
We'll get failures from these tests if ktls was enabled but was not available in this kernel version.

borisPis replied:
Should be fixed, see previous comment about runtime check.

mattcaswell replied:
Yes this works now.

borisPis commented on Configure:
```diff
@@ -322,6 +322,7 @@ my @dtls = qw(dtls1 dtls1_2);
 # For developers: keep it sorted alphabetically
 
 my @disablables = (
+    "ktls",
```
Done

borisPis commented on include/internal/bio.h:
```diff
@@ -31,3 +34,47 @@ void bio_cleanup(void);
+
+#ifdef  __cplusplus
+extern "C" {
+#endif
```
Done

borisPis commented on include/openssl/evp.h:
```diff
@@ -304,6 +304,7 @@ int (*EVP_CIPHER_meth_get_ctrl(const EVP_CIPHER *cipher))(EVP_CIPHER_CTX *,
 # define         EVP_CTRL_GCM_SET_TAG            EVP_CTRL_AEAD_SET_TAG
 # define         EVP_CTRL_GCM_SET_IV_FIXED       EVP_CTRL_AEAD_SET_IV_FIXED
 # define         EVP_CTRL_GCM_IV_GEN             0x13
+# define         EVP_CTRL_GCM_GET_IV             0x14
```
Done

borisPis commented on ssl/record/rec_layer_s3.c:
```diff
@@ -853,15 +865,19 @@ int do_ssl3_write(SSL *s, int type, const unsigned char *buf,
+         */
+        if (!BIO_get_ktls(s->wbio)
+                && (!WPACKET_put_bytes_u8(thispkt, rectype)
                 || !WPACKET_put_bytes_u16(thispkt, version)
```
Expanded the comment.
The version was set when kTLS was started:
`crypto_info.info.version = s->version;`

I think this code might need to be updated once ktls support is expanded to TLS1.3,. However, we will try to avoid it.

borisPis commented on ssl/t1_enc.c:
```diff
@@ -316,6 +325,64 @@ int tls1_change_cipher_state(SSL *s, int which)
         goto err;
     }
+#ifndef OPENSSL_NO_KTLS
+    if (s->compress || SSL_TREAT_AS_TLS13(s))
```
Oh, I've missed that.  Removing this check.

borisPis commented on test/sslapitest.c:
```diff
@@ -656,6 +657,178 @@ static int execute_test_large_message(const SSL_METHOD *smeth,
+    return testresult;
+}
+
+static int test_ktls_client_server()
```
Done.

borisPis commented on test/sslapitest.c:
```diff
@@ -5617,6 +5790,14 @@ int setup_tests(void)
+    ADD_TEST(test_ktls_client_server);
+    ADD_TEST(test_ktls_no_client_server);
+    ADD_TEST(test_ktls_client_no_server);
+    ADD_TEST(test_ktls_no_client_no_server);
```
Right, I've added a runtime check to prevent this from happening.

borisPis commented on include/internal/ktls.h:
```diff
@@ -0,0 +1,147 @@
+#    ifndef PEDANTIC
+#     warning "KTLS requires Kernel Headers >= 4.13.0"
+#     warning "Skipping Compilation of KTLS data path"
+#    endif
```
Assuming that it would be polite to print a warning and keep everything compiling properly. This would allow enabling it by default in the future without any complications.

borisPis commented on test/sslapitest.c:
```diff
@@ -656,6 +657,178 @@ static int execute_test_large_message(const SSL_METHOD *smeth,
+    /* Create a session based on SHA-256 */
+    if (!TEST_true(create_ssl_ctx_pair(TLS_server_method(),
+                                       TLS_client_method(),
+                                       TLS1_2_VERSION, TLS1_2_VERSION,
```
I've tried to avoid it in the ktls tests by enclosing them with `#ifndef OPENSSL_NO_TLS1_2`.

Are you sure it is still broken?

borisPis commented on test/sslapitest.c:
```diff
@@ -656,6 +657,178 @@ static int execute_test_large_message(const SSL_METHOD *smeth,
+        if (!TEST_false(BIO_get_ktls(clientssl->wbio)))
+            goto end;
+    } else {
+        if (!TEST_true(BIO_get_ktls(clientssl->wbio)))
```
Your kernel is probably recent enough, but ktls is not up. The run time check should fix that. Please re-run the test now.

mattcaswell replied:
Can you push your latest fixes?

borisPis replied:
Just did. Sorry about that.

mattcaswell commented:
After this:
````
$ ./config -d --strict-warnings enable-ktls no-tls1_2
$ make
````

I get:

````
gcc  -Iinclude -I. -pthread -m64 -Wa,--noexecstack -DDEBUG_UNUSED -DPEDANTIC -pedantic -Wno-long-long -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wswitch -Wsign-compare -Wmissing-prototypes -Wstrict-prototypes -Wshadow -Wformat -Wtype-limits -Wundef -Werror -Wall -O0 -g   -MMD -MF test/sslapitest-bin-sslapitest.d.tmp -MT test/sslapitest-bin-sslapitest.o -c -o test/sslapitest-bin-sslapitest.o test/sslapitest.c
test/sslapitest.c:838:12: error: 'test_ktls_no_client_no_server' defined but not used [-Werror=unused-function]
 static int test_ktls_no_client_no_server(void)
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
test/sslapitest.c:833:12: error: 'test_ktls_client_no_server' defined but not used [-Werror=unused-function]
 static int test_ktls_client_no_server(void)
            ^~~~~~~~~~~~~~~~~~~~~~~~~~
test/sslapitest.c:828:12: error: 'test_ktls_no_client_server' defined but not used [-Werror=unused-function]
 static int test_ktls_no_client_server(void)
            ^~~~~~~~~~~~~~~~~~~~~~~~~~
test/sslapitest.c:823:12: error: 'test_ktls_client_server' defined but not used [-Werror=unused-function]
 static int test_ktls_client_server(void)
            ^~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
Makefile:14938: recipe for target 'test/sslapitest-bin-sslapitest.o' failed
````

mattcaswell commented on ssl/t1_enc.c:
```diff
@@ -316,6 +325,63 @@ int tls1_change_cipher_state(SSL *s, int which)
+
+    /* ktls doesn't support renegotiation */
+    if (BIO_get_ktls(s->wbio))
+        goto err;
```
I tried a renegotiation from s_client to see what would happen. I got an assertion failed error and abort! This error needs to call `SSLfatal` first. We later do a sanity check to make sure we've sent an alert via SSLfatal if we're in an error state. That assertion was failing in this case (also see #7594). I wonder if we should automatically set `SSL_OP_NO_RENEGOTIATION` if ktls is in use.

mattcaswell replied:
...or possibly explicitly check if we're using ktls everywhere we check `SSL_OP_NO_RENEGOTIATION`

borisPis replied:
Fixed the compilation issue.
Added the call to SSLFatal, and set the SSL_OP_NO_RENEGOTATION option on the SSL object.

borisPis commented:
Ping @mattcaswell. AFAIK all the changes requested were made. Could you please take another look.

mattcaswell commented on test/sslapitest.c:
```diff
@@ -656,6 +657,194 @@ static int execute_test_large_message(const SSL_METHOD *smeth,
 }
 
+#ifndef OPENSSL_NO_TLS1_2
+#ifndef OPENSSL_NO_KTLS
```
Probably this would be better on one line `#if !defined(OPENSSL_NO_TLS1_2) && !defined(OPENSSL_NO_KTLS)`
Alternatively, if done on two lines then the second one receives a single space indent:
````
#ifndef OPENSSL_NO_TLS1_2
# ifndef OPENSSL_NO_KTLS
````
And then similarly with the endifs:
````
# endif
#endif
````

mattcaswell commented:
Oh. I forgot to mention: "git log -p" reveals a few lines with trailing whitespace. It would be nice to clean that up.

borisPis commented on doc/man3/SSL_CTX_set_mode.pod:
```diff
@@ -105,6 +105,21 @@ Enable asynchronous processing. TLS I/O operations may indicate a retry with
+Disable the use of the kernel TLS egress data-path.
+By default kernel TLS is enabled if it is supported by the negotiated ciphersuites
+and extensions. The kernel TLS data-path implements the record layer,
+and the crypto algorithm. The kernel will utilize the best HW
```
Fixed.

borisPis commented on doc/man3/SSL_CTX_set_mode.pod:
```diff
@@ -105,6 +105,21 @@ Enable asynchronous processing. TLS I/O operations may indicate a retry with
+
+Disable the use of the kernel TLS egress data-path.
+By default kernel TLS is enabled if it is supported by the negotiated ciphersuites
+and extensions. The kernel TLS data-path implements the record layer,
```
Sure.

borisPis commented on crypto/bio/bss_sock.c:
```diff
@@ -108,10 +120,20 @@ static int sock_read(BIO *b, char *out, int outl)
-    ret = writesocket(b->num, in, inl);
+# ifndef OPENSSL_NO_KTLS
+    if (BIO_should_ktls_ctrl_msg_flag(b)) {
+        unsigned char record_type = (intptr_t) b->ptr;
```
Sure.

borisPis commented on INSTALL:
```diff
@@ -250,6 +250,15 @@
+                   performance and allow for the use of sendfile and splice
+                   system calls on TLS sockets. The Kernel may use TLS
+                   accelerators if any are available on the system.
+                   This option will be forced on systems that do not support
```
Yes. Fixed.

borisPis commented:
We can find out whether ktls is in use via BIO_get_ktls. But, you recently requested to move it into the private header... I'll move it back to the public header and add the changes in s_client and s_server.

borisPis commented:
All fixes are pushed. @mattcaswell please take another look.

mattcaswell commented:
> We can find out whether ktls is in use via BIO_get_ktls. But, you recently requested to move it into the private header... I'll move it back to the public header and add the changes in s_client and s_server.

I'm fine with things being in the public header if there is a need for them in the public API. What I don't like is essentially private implementation details becoming part of the public API by accident. We have had a lot of that in the past simply because we weren't careful enough about clearly defining our public API.

mattcaswell commented on CHANGES:
```diff
@@ -34,6 +34,10 @@
 
+  *) Added support for Linux Kernel TLS data-path. The Linux Kernel data-path
+     improves application performance by removing data copies and providing
+     applications with zero-copy system calls such as sendfile and splice.
```
Nit: Please add your name after this entry, as per the other entries in this file

borisPis replied:
Done

mattcaswell commented on apps/s_client.c:
```diff
@@ -3245,6 +3245,12 @@ static void print_stuff(BIO *bio, SSL *s, int full)
+    if (BIO_get_ktls(SSL_get_wbio(s)))
+        BIO_printf(bio_err, "Using Kernel TLS for sending\n");
+    if (BIO_get_ktls(SSL_get_rbio(s)))
+        BIO_printf(bio_err, "Using Kernel TLS for receiving\n");
```
Actually - that raises an API question. Do we actually need something like `BIO_get_ktls_recv()` and `BIO_get_ktls_send()` or similar? Or (perhaps) just the latter for now since we don't (yet) support the former.

borisPis replied:
We do need it for both directions eventually. I think that my next patchset that introduces receive support for kTLS in OpenSSL should add that. I'll remove the receive side print, because it doesn't make any sense for the code at the moment.

borisPis commented:
Ping @mattcaswell all fixes are now pushed to github.

mattcaswell commented on doc/man3/BIO_ctrl.pod:
```diff
@@ -72,6 +74,9 @@ Not all BIOs support these calls. BIO_ctrl_pending() and BIO_ctrl_wpending()
 return a size_t type and are functions, BIO_pending() and BIO_wpending() are
 macros which call BIO_ctrl().
 
+BIO_get_ktls() return 1 if the BIO is using the Kernel TLS data-path for
```
Would `BIO_get_ktls_send()` or similar be a better name? Given that we know we are going to need a similar macro for the receive side int he future

borisPis replied:
Done.

mattcaswell commented on doc/man3/BIO_ctrl.pod:
```diff
@@ -34,6 +34,8 @@ BIO_get_info_callback, BIO_set_info_callback, BIO_info_cb
  int BIO_get_info_callback(BIO *b, BIO_info_cb **cbp);
  int BIO_set_info_callback(BIO *b, BIO_info_cb *cb);
 
+ int BIO_get_ktls(BIO *b);
```
Nit: Please add a HISTORY section in this man page for this macro (see other man pages for an example)

borisPis replied:
Done.

mattcaswell commented on doc/man3/SSL_CTX_set_mode.pod:
```diff
@@ -105,6 +105,22 @@ Enable asynchronous processing. TLS I/O operations may indicate a retry with
 SSL_ERROR_WANT_ASYNC with this mode set if an asynchronous capable engine is
 used to perform cryptographic operations. See L<SSL_get_error(3)>.
 
+=item SSL_MODE_NO_KTLS_TX
```
Nit: Please add this to the HISTORY section below.

borisPis replied:
Done

FdaSilvaYY commented on include/openssl/bio.h:
```diff
@@ -145,6 +145,16 @@ extern "C" {
+
+#  define BIO_CTRL_GET_KTLS                 73
+#  define BIO_get_ktls_send(b)         \
+     BIO_ctrl(b, BIO_CTRL_GET_KTLS, 0, NULL)
```
rename BIO_CTRL_GET_KTLS as BIO_get_ktls_send ?

borisPis replied:
It must be capital case as all other defines.

I'll adjust the name to make the direction more clear in the define.

mattcaswell commented:
Ping @openssl/committers. Still looking for a second reviewer for this.

@borisPis - Please can you remove the last "merge" commit from this PR, and instead rebase it (if necessary). We don't allow merge commits.

borisPis commented:
Thanks @mattcaswell, I've removed the merge commit and rebased the changes.

@openssl we are still looking for a second reviewer for this. Please take a look.

I'd rebase and push the receive side support for KTLS once this is accepted.

borisPis commented:
Ping @levitte, all done

InfoHunter commented on doc/man3/BIO_ctrl.pod:
```diff
@@ -124,6 +132,10 @@ particular a return value of 0 can be returned if an operation is not
 
+=head1 HISTORY
+
+The BIO_get_ktls_send() function was added in OpenSSL 1.1.1.
```
Or `was added in OpenSSL 3.0.0.`...

mspncp commented on doc/man3/SSL_CTX_set_mode.pod:
```diff
@@ -125,6 +141,7 @@ L<SSL_write(3)>, L<SSL_get_error(3)>
 =head1 HISTORY
 
 SSL_MODE_ASYNC was first added to OpenSSL 1.1.0.
+SSL_MODE_NO_KTLS_TX was first added to OpenSSL 1.1.1.
```
FWIW: A grep yields that the standard formulation is "This function was added in OpenSSL x.x.x" (with exceptions).

InfoHunter replied:
> A PR to fix the exceptions would be welcome.

Absolutely

mspncp replied:
@InfoHunter you you'd want to do it, go ahead. If not, I can do it, too.

InfoHunter replied:
I expect you can do this @mspncp since I don't have enough time recently, sorry :-(

InfoHunter commented:
It seems this could be considered 'ready' and enters a 'grace' period
